courses/models.py

```diff
@@ -301,7 +302,33 @@ def courses(self):
         - list of tuple (Course, string): courses that are either requirements or electives, plus the requirement type
         """
 
-        return self._req_course_walk(self.requirements_root, [])
+        heap = []
+        main_ops = ProgramRequirement.objects.filter(program=self, depth=2).all()
+
+        for op in main_ops:
+            reqs = (
+                ProgramRequirement.objects.filter(
+                    program__id=self.id,
+                    path__startswith=op.path,
+                    node_type=ProgramRequirementNodeType.COURSE,
+                )
+                .select_related("course", "course__page")
+                .all()
+            )
+
+            heap.extend(
+                [
+                    (
+                        req.course,
+                        "Required Courses"
+                        if not op.elective_flag
```

Comment:
Can the requirement operator be used instead of elective_flag to determine whether the course is a requirement or elective? `ProgramRequirement.Operator.MIN_NUMBER_OF.value` vs `ProgramRequirement.Operator.ALL_OF.value`?

Reply:
It depends. Having the flag here allows for electives to also be "all of" at its root, and I can think of a scenario where that would be needed: the electives are grouped further into two or more "min number of" groups. 
```
Elective ---------> Group 1       ---------> Course 1
ALL_OF     \        MIN_NUMBER_OF   \ -----> Course 2
            \
             \----> Group 2       ---------> Course 3
                    MIN_NUMBER_OF   \ -----> Course 4
```

In this case, both second-level nodes would have the same operator, so telling them apart programmatically isn't possible (without the flag).

Reply:
Got it, good thinking and nice diagram!